src/psyclone/docstring_parser.py

```diff
@@ -0,0 +1,539 @@
+                    f"'{param}' type found with no parameter docstring."
+                )
+            docstring_data.arguments[param].datatype = types[param]
+            docstring_data.arguments[param].inline_type = False
+
+        # Store the return type if specified.
+        if rtype:
+            docstring_data.returns.datatype = rtype
+            docstring_data.returns.inline_type = False
+
+        return docstring_data
+
+
+def create_docstring_data(args: List[str], desc: str, obj: Any) ->\
+        Union[ArgumentData, RaisesData, ReturnsData]:
+    '''
+    Creates a docstring data object (i.e. ArgumentData, RaisesData or
+    ReturnsData) from an argument list and description.
+
+    :param args: The argument list (e.g. ['param', 'args']) for the docstring
+                 entry.
+    :param desc: The description corresponding to the docstring element.
+    :param obj: The object that is having its docstring analysed.
+
+    :raises DocParseError: If an unsupported docstring input is found.
+
+    :returns: An object representing the input arg and desc.
+    '''
+    # If its a param then we can create an ArgumentData for this.
+    if args[0] == "param":
+        if len(args) == 2:
+            # Parameter of form :param name: desc.
+            inline_type = False
+            datatype = None
+            name = args[1]
+            # Check if there is a datatype on the argument itself and store
+            # it.
+            signature = inspect.signature(obj)
+            val = signature.parameters.get(name, None)
+            if (val is not None and val.annotation
+                    is not inspect.Parameter.empty):
+                datatype = stringify_annotation(
+                    val.annotation
+                )
+                inline_type = True
+        elif len(args) == 3:
+            # Parameter of form :param type name: desc.
+            inline_type = True
+            datatype = args[1]
+            name = args[2]
+        else:
+            raise DocParseError(
+                f"Found parameter docstring of unsupported type, expected "
+                f":param arg: or :param type arg: but found "
+                f":{' '.join(args)}:"
+            )
+        return ArgumentData(
+            name=name, desc=desc, datatype=datatype, inline_type=inline_type
+        )
+    elif args[0] == "raises":
+        # Raises must be of form :raises Error: desc
+        if len(args) != 2:
+            raise DocParseError(
+                f"Found raises docstring of unsupported type, expected "
+                f":raises Error: but found :{' '.join(args)}:"
+            )
+        return RaisesData(exception=args[1], desc=desc)
+    elif args[0] == "returns":
+        if len(args) == 1:
+            # Return of form :returns: desc
+            inline_type = False
+            datatype = None
+        elif len(args) == 2:
+            # Return of form :returns datatype: desc
```

Comment:
As mentioned earlier - are you sure this is valid Sphinx syntax?

Reply:
I had made an assumption but I'm not sure it is so I've removed it. If we don't use it thats good enough for us.